File: poc/ts/src/walk.ts
import { readdirSync } from "node:fs";
import { join } from "node:path";

/** Recursive file walk. Hand-rolled to mirror Rust's walkdir for a fair comparison. */
export function* walk(dir: string): Generator<string> {
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return; // missing or unreadable directory is not fatal
  }
  for (const e of entries) {
    const p = join(dir, e.name);
    if (e.isDirectory()) yield* walk(p);
    else if (e.isFile()) yield p;
  }
}

File: poc/rust/src/walk.rs
use std::path::{Path, PathBuf};

/// Recursive file walk. Hand-rolled to mirror the TS implementation exactly.
pub fn walk(dir: &Path, out: &mut Vec<PathBuf>) {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return; // missing or unreadable directory is not fatal
    };
    for entry in entries.flatten() {
        let path = entry.path();
        match entry.file_type() {
            Ok(ft) if ft.is_dir() => walk(&path, out),
            Ok(ft) if ft.is_file() => out.push(path),
            _ => {}
        }
    }
}
